esp_linenoise/include/esp_linenoise.h

```diff
@@ -0,0 +1,303 @@
+    int in_fd; /*!< File descriptor to read input from (e.g., STDIN_FILENO) */
+    int out_fd; /*!< File descriptor to write output to (e.g., STDOUT_FILENO) */
+    bool allow_multi_line; /*!< Whether to allow multi-line input (true to enable) */
+    bool allow_empty_line; /*!< Whether to allow accepting an empty line as valid input */
+    bool allow_dumb_mode; /*!< Whether to allow running in dumb terminal mode (without advanced features) */
+    esp_linenoise_completion_callback completion_cb; /*!< Callback function for handling input completions */
+    esp_linenoise_hints_callback hints_cb; /*!< Callback function to provide input hints (e.g., usage suggestions) */
+    esp_linenoise_free_hints_callback free_hints_cb; /*!< Callback function to free hints returned by `hints_cb` */
+    void *user_ctx; /*!< User context information passed to the callbacks */
+    esp_linenoise_read_bytes_fn read_bytes_fn; /*!< Function used to read bytes from the input stream */
+    esp_linenoise_write_bytes_fn write_bytes_fn; /*!< Function used to write bytes to the output stream */
+    char **history; /*!< Pointer to the history buffer (used internally; typically initialized to NULL) */
+} esp_linenoise_config_t;
+
+/**
+ * @brief Opaque handle to a linenoise instance.
+ */
+typedef struct esp_linenoise_instance *esp_linenoise_handle_t;
+
+/**
+ * @brief Returns the default parameters for creating a linenoise instance.
+ *
+ * @param config esp_linenoise_config_t structure to populate with default values.
+ */
+void esp_linenoise_get_instance_config_default(esp_linenoise_config_t *config);
+
+/**
+ * @brief Creates a new linenoise instance.
+ *
+ * @param param Pointer to the configuration parameters for the instance.
+ * @return Handle to the created instance, or NULL on failure.
+ */
+esp_linenoise_handle_t esp_linenoise_create_instance(esp_linenoise_config_t *param);
+
+/**
+ * @brief Destroys a linenoise instance and frees associated memory.
+ *
+ * @param handle Handle to the instance to delete.
+ * @return ESP_OK on success, or error code on failure.
+ */
+esp_err_t esp_linenoise_delete_instance(esp_linenoise_handle_t handle);
+
+/**
+ * @brief This function is used by the callback function registered by the user
+ * in order to add completion options given the input string when the
+ * user typed <tab>. See the example.c source code for a very easy to
+ * understand example.
+ *
+ * @param lc line completion structure being filled be the function
+ * @param str command to complete
+ */
+void esp_linenoise_add_completion(esp_linenoise_completions_t *lc, const char *str);
+
+/**
+ * @brief Reads a line of input from the user into a buffer.
+ *
+ * @param handle Handle to the linenoise instance.
+ * @param cmd_line_buffer Buffer to store the input line.
+ * @param cmd_line_length Length of the command line buffer.
+ * @return ESP_OK on success, or error code on failure.
```

Comment:
It would be nice to clarify what happens if cmd_line_buffer is not large enough to fit a line of input. Do we return ESP_OK or an error, is the line truncated, etc.